checks.d/spark.py

```diff
@@ -0,0 +1,457 @@
+
+        # Get the rdd metrics
+        self._spark_rdd_metrics(spark_apps, tags)
+
+        # Report success after gathering all metrics from the ApplicationMaster
+        if running_apps:
+            app_id, (app_name, tracking_url) = running_apps.items()[0]
+            am_address = self._get_url_base(tracking_url)
+
+            self.service_check(SPARK_SERVICE_CHECK,
+                AgentCheck.OK,
+                tags=['url:%s' % am_address],
+                message='Connection to ApplicationMaster "%s" was successful' % am_address)
+
+    def _get_running_spark_apps(self, rm_address):
+        '''
+        Return a dictionary of {app_id: (app_name, tracking_url)} for the running Spark applications
+        '''
+        metrics_json = self._rest_request_to_json(rm_address,
+            YARN_APPS_PATH,
+            YARN_SERVICE_CHECK,
+            states=YARN_APPLICATION_STATES,
+            applicationTypes=YARN_APPLICATION_TYPES)
+
+        running_apps = {}
+
+        if metrics_json.get('apps'):
+            if metrics_json['apps'].get('app') is not None:
+
+                for app_json in metrics_json['apps']['app']:
+                    app_id = app_json.get('id')
+                    tracking_url = app_json.get('trackingUrl')
+                    app_name = app_json.get('name')
+
+                    if app_id and tracking_url and app_name:
+                        running_apps[app_id] = (app_name, tracking_url)
+
+        return running_apps
+
+    def _get_spark_app_ids(self, running_apps):
+        '''
+        Return a dictionary of {app_id: (app_name, tracking_url)} for Spark applications
+        '''
+        spark_apps = {}
+        for app_id, (app_name, tracking_url) in running_apps.iteritems():
+            response = self._rest_request_to_json(tracking_url,
+                SPARK_APPS_PATH,
+                SPARK_SERVICE_CHECK)
+            for app in response:
+                app_id = app.get('id')
+                app_name = app.get('name')
+
+                if app_id and app_name:
+                    spark_apps[app_id] = (app_name, tracking_url)
+
+        return spark_apps
+
+    def _spark_job_metrics(self, running_apps, addl_tags):
+        '''
+        Get metrics for each Spark job.
+        Return a map from Stage IDs to Job IDs
+        '''
+        for app_id, (app_name, tracking_url) in running_apps.iteritems():
+
+            response = self._rest_request_to_json(tracking_url,
+                SPARK_APPS_PATH,
+                SPARK_SERVICE_CHECK, app_id, 'jobs')
+
+            for job in response:
+
+                status = job.get('status')
+
+                tags = ['app_name:%s' % str(app_name)]
+                tags.extend(addl_tags)
+                tags.append('status:%s' % str(status).lower())
+
+                self._set_metrics_from_json(tags, job, SPARK_JOB_METRICS)
+
+            if len(response):
+                self._set_metric('spark.job.count', INCREMENT, len(response), tags)
```

Comment:
The `tags` that are passed here are the `tags` of the last job of the loop above, so the `status` tag wouldn't be accurate.

Instead we should either:
- increment this count by 1 for every job, and with the `status` tag (in order to have a `count` per status)
- increment this count by `len(response)` only once per app, and without the `status` tag (in order to have only a global `count` for all status)

I would go with option 1 since having a count per status sounds useful to me, but let me know what you think.

This applies to the `spark.stage.count` metric too.

Also, if possible it'd be good to have tests that reflect how different statuses on the `jobs` and `stages` should be counted.